File: python/src/utils/debug_helper.py
"""
Debug Helper
Provides debugging utilities for Printify API interactions
"""

from typing import List, Dict, Any
from services.printify_api import PrintifyApiClient


class DebugHelper:
    """Helper class for debugging Printify API interactions"""
    
    def __init__(self, api_client: PrintifyApiClient):
        self.api_client = api_client
    
    def debug_blueprints(self) -> None:
        """Debug blueprint information"""
        try:
            print("🔍 Fetching blueprints...")
            blueprints = self.api_client.get_blueprints()
            
            print(f"✅ Found {len(blueprints)} blueprints:")
            for i, blueprint in enumerate(blueprints[:10], 1):
                print(f"  {i}. ID: {blueprint.id} - {blueprint.title}")
                print(f"     Brand: {blueprint.brand}, Model: {blueprint.model}")
                print(f"     Description: {blueprint.description[:100]}...")
                print()
            
            if len(blueprints) > 10:
                print(f"... and {len(blueprints) - 10} more blueprints")
        except Exception as e:
            print(f"❌ Error fetching blueprints: {e}")
    
    def debug_print_providers(self, blueprint_id: int) -> None:
        """Debug print providers for a specific blueprint"""
        try:
            print(f"🔍 Fetching print providers for blueprint {blueprint_id}...")
            providers = self.api_client.get_print_providers(blueprint_id)
            
            print(f"✅ Found {len(providers)} print providers:")
            for i, provider in enumerate(providers, 1):
                print(f"  {i}. ID: {provider.id} - {provider.title}")
                print(f"     Location: {provider.location}")
                print()
        except Exception as e:
            print(f"❌ Error fetching print providers for blueprint {blueprint_id}: {e}")
    
    def debug_variants(self, blueprint_id: int, print_provider_id: int) -> None:
        """Debug variants for a specific blueprint and print provider"""
        try:
            print(f"🔍 Fetching variants for blueprint {blueprint_id}, print provider {print_provider_id}...")
            response = self.api_client.get_variants(blueprint_id, print_provider_id)
            
            # Handle the actual response structure
            variants = response.get('variants', response) if isinstance(response, dict) else response
            
            print(f"✅ Found {len(variants) if variants else 0} variants:")
            
            if variants and len(variants) > 0:
                for i, variant in enumerate(variants[:5], 1):
                    print(f"  {i}. ID: {variant.get('id')} - {variant.get('title', 'Untitled')}")
                    print(f"     Options: {variant.get('options', 'No options')}")
                    print(f"     Placeholders: {len(variant.get('placeholders', []))} positions available")
                    print(f"     Decoration Methods: {', '.join(variant.get('decoration_methods', []))}")
                    print()
                
                if len(variants) > 5:
                    print(f"... and {len(variants) - 5} more variants")
            else:
                print("No variants found or invalid response format")
                print(f"Raw response: {response}")
        except Exception as e:
            print(f"❌ Error fetching variants for blueprint {blueprint_id}, print provider {print_provider_id}: {e}")
    
    def debug_blueprint_complete(self, blueprint_id: int) -> None:
        """Debug all information for a specific blueprint"""
        print(f"🔍 Complete debug for blueprint {blueprint_id}")
        print("=" * 50)
        
        self.debug_print_providers(blueprint_id)
        
        # Get print providers first
        try:
            providers = self.api_client.get_print_providers(blueprint_id)
            if providers:
                first_provider = providers[0]
                print(f"\n🔍 Using first print provider: {first_provider.title} (ID: {first_provider.id})")
                self.debug_variants(blueprint_id, first_provider.id)
        except Exception as e:
            print(f"❌ Error in complete debug: {e}")
    
    def debug_print_provider(self, provider_id: int) -> None:
        """Debug a specific print provider by ID"""
        try:
            print(f"🔍 Fetching print provider with ID: {provider_id}...")
            provider = self.api_client.get_print_provider(provider_id)
            
            print(f"✅ Print Provider Details:")
            print(f"  ID: {provider.id}")
            print(f"  Title: {provider.title}")
            print(f"  Location: {provider.location}")
            
            # Note: Print provider API doesn't return variants directly
            # To get variants, we need to know which blueprint this provider works with
            print(f"  Note: To see variants for this provider, use:")
            print(f"    python main.py debug-structure <blueprint_id> {provider.id}")
        except Exception as e:
            print(f"❌ Error fetching print provider {provider_id}: {e}")
    
    def show_recommended_product_structure(self, blueprint_id: int, print_provider_id: int) -> None:
        """Show recommended product structure based on actual data"""
        try:
            print("🔍 Generating recommended product structure...")
            
            response = self.api_client.get_variants(blueprint_id, print_provider_id)
            variants = response.get('variants', response) if isinstance(response, dict) else response
            
            if not variants or len(variants) == 0:
                print("❌ No variants found for this blueprint/print provider combination")
                print(f"Raw response: {response}")
                return
            
            first_variant = variants[0]
            print("\n📋 Recommended product.json structure:")
            
            # Get blueprint info
            blueprints = self.api_client.get_blueprints()
            blueprint = next((bp for bp in blueprints if bp.id == blueprint_id), None)
            blueprint_title = blueprint.title if blueprint else "Your Product"
            
            # Get provider info
            providers = self.api_client.get_print_providers(blueprint_id)
            provider = next((p for p in providers if p.id == print_provider_id), None)
            provider_title = provider.title if provider else "Your Provider"
            
            recommended_structure = {
                "title": "Your Product Title",
                "description": "Your product description",
                "blueprint_id": blueprint_id,
                "print_provider_id": print_provider_id,
                "variants": [
                    {
                        "id": first_variant.get('id'),
                        "price": 2500,  # $25.00 in cents
                        "is_enabled": True,
                        "is_default": True,
                        "grams": 180,  # Default weight
                        "options": first_variant.get('options', [])
                    }
                ],
                "print_areas": [
                    {
                        "variant_ids": [first_variant.get('id')],
                        "placeholders": [
                            {
                                "position": "front",
                                "images": [
                                    {
                                        "id": "your_image_id",
                                        "name": "Your Design Name",
                                        "url": "https://your-image-url.com/image.png",
                                        "preview_url": "https://your-image-url.com/image.png",
                                        "x": 0,
                                        "y": 0,
                                        "scale": 1,
                                        "angle": 0
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }
            
            import json
            print(json.dumps(recommended_structure, indent=2))
            
            print("\n📏 Available print positions for this variant:")
            if first_variant.get('placeholders'):
                for i, placeholder in enumerate(first_variant['placeholders'], 1):
                    print(f"  {i}. {placeholder.get('position')} ({placeholder.get('width', 'N/A')}x{placeholder.get('height', 'N/A')})")
        except Exception as e:
            print(f"❌ Error generating recommended structure: {e}") 

File: src/utils/debugHelper.ts
import { PrintifyApiClient } from "../services/printifyApi";

export class DebugHelper {
  private apiClient: PrintifyApiClient;

  constructor(apiClient: PrintifyApiClient) {
    this.apiClient = apiClient;
  }

  /**
   * Debug blueprint information
   */
  async debugBlueprints(): Promise<void> {
    try {
      console.log("🔍 Fetching blueprints...");
      const blueprints = await this.apiClient.getBlueprints();

      console.log(`✅ Found ${blueprints.length} blueprints:`);
      blueprints.slice(0, 10).forEach((blueprint, index) => {
        console.log(`  ${index + 1}. ID: ${blueprint.id} - ${blueprint.title}`);
        console.log(`     Brand: ${blueprint.brand}, Model: ${blueprint.model}`);
        console.log(`     Description: ${blueprint.description.substring(0, 100)}...`);
        console.log("");
      });

      if (blueprints.length > 10) {
        console.log(`... and ${blueprints.length - 10} more blueprints`);
      }
    } catch (error) {
      console.error("❌ Error fetching blueprints:", error);
    }
  }

  /**
   * Debug print providers for a specific blueprint
   */
  async debugPrintProviders(blueprintId: number): Promise<void> {
    try {
      console.log(`🔍 Fetching print providers for blueprint ${blueprintId}...`);
      const providers = await this.apiClient.getPrintProviders(blueprintId);

      console.log(`✅ Found ${providers.length} print providers:`);
      providers.forEach((provider, index) => {
        console.log(`  ${index + 1}. ID: ${provider.id} - ${provider.title}`);
        console.log(`     Location: ${provider.location}`);
        console.log("");
      });
    } catch (error) {
      console.error(`❌ Error fetching print providers for blueprint ${blueprintId}:`, error);
    }
  }

  /**
   * Debug a specific print provider by ID
   */
  async debugPrintProvider(providerId: number): Promise<void> {
    try {
      console.log(`🔍 Fetching print provider with ID: ${providerId}...`);
      const provider = await this.apiClient.getPrintProvider(providerId);

      console.log(`✅ Print Provider Details:`);
      console.log(`  ID: ${provider.id}`);
      console.log(`  Title: ${provider.title}`);
      console.log(`  Location: ${typeof provider.location === "object" ? JSON.stringify(provider.location) : provider.location}`);

      // Note: Print provider API doesn't return variants directly
      // To get variants, we need to know which blueprint this provider works with
      console.log(`  Note: To see variants for this provider, use:`);
      console.log(`    yarn start debug-structure <blueprint_id> ${provider.id}`);
    } catch (error) {
      console.error(`❌ Error fetching print provider ${providerId}:`, error);
    }
  }

  /**
   * Debug variants for a specific blueprint and print provider
   */
  async debugVariants(blueprintId: number, printProviderId: number): Promise<void> {
    try {
      console.log(`🔍 Fetching variants for blueprint ${blueprintId}, print provider ${printProviderId}...`);
      const response = await this.apiClient.getVariants(blueprintId, printProviderId);

      // Handle the actual response structure
      const variants = (response as any)?.variants || response;

      console.log(`✅ Found ${variants?.length || 0} variants:`);

      if (Array.isArray(variants) && variants.length > 0) {
        variants.slice(0, 5).forEach((variant, index) => {
          console.log(`  ${index + 1}. ID: ${variant.id} - ${variant.title}`);
          console.log(`     Options:`, variant.options || "No options");
          console.log(`     Placeholders: ${variant.placeholders?.length || 0} positions available`);
          console.log(`     Decoration Methods: ${variant.decoration_methods?.join(", ") || "None"}`);
          console.log("");
        });

        if (variants.length > 5) {
          console.log(`... and ${variants.length - 5} more variants`);
        }
      } else {
        console.log("No variants found or invalid response format");
        console.log("Raw response:", JSON.stringify(response, null, 2));
      }
    } catch (error) {
      console.error(`❌ Error fetching variants for blueprint ${blueprintId}, print provider ${printProviderId}:`, error);
    }
  }

  /**
   * Debug all information for a specific blueprint
   */
  async debugBlueprintComplete(blueprintId: number): Promise<void> {
    console.log(`🔍 Complete debug for blueprint ${blueprintId}`);
    console.log("=".repeat(50));

    await this.debugPrintProviders(blueprintId);

    // Get print providers first
    try {
      const providers = await this.apiClient.getPrintProviders(blueprintId);
      if (providers.length > 0) {
        const firstProvider = providers[0];
        console.log(`\n🔍 Using first print provider: ${firstProvider.title} (ID: ${firstProvider.id})`);
        await this.debugVariants(blueprintId, firstProvider.id);
      }
    } catch (error) {
      console.error("❌ Error in complete debug:", error);
    }
  }

  /**
   * Show recommended product structure based on actual data
   */
  async showRecommendedProductStructure(blueprintId: number, printProviderId: number): Promise<void> {
    try {
      console.log("🔍 Generating recommended product structure...");

      const response = await this.apiClient.getVariants(blueprintId, printProviderId);
      const variants = (response as any)?.variants || response;

      if (!Array.isArray(variants) || variants.length === 0) {
        console.log("❌ No variants found for this blueprint/print provider combination");
        console.log("Raw response:", JSON.stringify(response, null, 2));
        return;
      }

      const firstVariant = variants[0];
      console.log("\n📋 Recommended product.json structure:");
      console.log(
        JSON.stringify(
          {
            title: "Your Product Title",
            description: "Your product description",
            blueprint_id: blueprintId,
            print_provider_id: printProviderId,
            variants: [
              {
                id: firstVariant.id,
                price: 2500, // $25.00 in cents
                is_enabled: true,
                is_default: true,
                grams: 180, // Default weight
                options: firstVariant.options || [],
              },
            ],
            print_areas: [
              {
                variant_ids: [firstVariant.id],
                placeholders: [
                  {
                    position: "front",
                    images: [
                      {
                        id: "your_image_id",
                        name: "Your Design Name",
                        url: "https://your-image-url.com/image.png",
                        preview_url: "https://your-image-url.com/image.png",
                      },
                    ],
                  },
                ],
              },
            ],
          },
          null,
          2
        )
      );

      console.log("\n📏 Available print positions for this variant:");
      if (firstVariant.placeholders) {
        (firstVariant.placeholders as any[]).forEach((placeholder: any, index: number) => {
          console.log(`  ${index + 1}. ${placeholder.position} (${placeholder.width}x${placeholder.height})`);
        });
      }
    } catch (error) {
      console.error("❌ Error generating recommended structure:", error);
    }
  }
}
